### PR TITLE
Dummy PR for 'flow-dependent, cross-timescale model diagnostic' POD

### PR DESCRIPTION
**Pull request by drewmresnick**
Submission of Dummy PR

This PR only includes a partially complete diagnostics documentation file. The rest of the documentation and additional files are still being compiled.

### CODE REVIEW

File: diagnostics/flow_dep_diags/doc/flow_dep_diags.rst [hunk 1]
```diff
@@ -0,0 +1,136 @@
+.. This is a comment in RestructuredText format (two periods and a space).
```
Comment (jkrasting) [inline diagnostics/flow_dep_diags/doc/flow_dep_diags.rst:1]: The commented sections can be removed from this file.

**Review by jkrasting — CHANGES_REQUESTED** [review]
Thanks @drewmresnick

File: diagnostics/flow_dep_diags/doc/flow_dep_diags.rst (hunk 1 above)
Comment (drewmresnick) [reply diagnostics/flow_dep_diags/doc/flow_dep_diags.rst:1]: @jkrasting could you point me to who I should contact regarding submission of the data for this POD?

**Comment by bitterbark** [comment]
That would be me. Instructions here:
https://github.com/NOAA-GFDL/MDTF-diagnostics/discussions/125

**Review by tsjackson-noaa — APPROVED** [review]

**Review by tsjackson-noaa — APPROVED** [review]

File: diagnostics/flow_dep_diag/ClimAnom_func.py [hunk 2]
```diff
@@ -3,26 +3,22 @@
...
+    ds[time_var] = ds.indexes[time_var].to_datetimeindex() #convert time to datetime so we can use groupby functionality
     # Drop 1 dimensional coordinates
     ds = ds.squeeze()
-    da = ds[va]
+    da = ds[var_name]
     da_ensmean = da.copy()
-    #would need to make envt var or something so you could use either 'time' or 'S' as coordinate name without an issue
-    da_day_clim = da_ensmean.groupby('{time_coord}.dayofyear').mean(time_var).format(**os.environ)
-    da_day_anom = da.groupby('{time_coord}.dayofyear').format(**os.environ) - da_day_clim
+    da_day_clim = da_ensmean.groupby('{time_coord}.dayofyear'.format(**os.environ)).mean(time_var) #.format(**os.environ)
```
Comment (wrongkindofdoctor) [inline diagnostics/flow_dep_diag/ClimAnom_func.py:20]: is it necessary tto keep the `#.format(**os.environ)` comment?

File: diagnostics/flow_dep_diag/cropping.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import xarray as xr
+
+file_path =
```
Comment (wrongkindofdoctor) [inline diagnostics/flow_dep_diag/cropping.py:3]: set to None

File: diagnostics/flow_dep_diag/doc/flow_dep_diag.rst [hunk 4]
```diff
@@ -1,50 +1,157 @@
...
+.. code-block:: restructuredtext
+
+   .. _my-figure-tag: [only needed for linking to figures]
+
+   .. figure:: [path to image file, relative to the source.rst file]
+      :align: left
+      :width: 75 % [these both need to be indented by three spaces]
+
+      Paragraphs or other text following the figure that are indented by three
+      spaces are treated as a caption/legend, eg:
```
Comment (wrongkindofdoctor) [inline diagnostics/flow_dep_diag/doc/flow_dep_diag.rst:114]: Remove template lines

File: diagnostics/flow_dep_diag/flow_dep_diag.yml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  - typing_extensions=3.7.4.3=pyha847dfd_0
+  - wheel=0.36.2=pyhd3eb1b0_0
+  - xarray=0.17.0=pyhd3eb1b0_0
+  - xz=5.2.5=h1de35cc_0
+  - yaml=0.2.5=haf1e3a3_0
+  - zict=2.0.0=pyhd3eb1b0_0
+  - zipp=3.4.1=pyhd3eb1b0_0
+  - zlib=1.2.11=h1de35cc_3
+  - zstd=1.4.9=h322a384_0
+prefix: /Users/drewr/opt/anaconda3/envs/MDTF_flow_dep_diag
```
Comment (wrongkindofdoctor) [inline diagnostics/flow_dep_diag/flow_dep_diag.yml:104]: remove prefix definition

File: data/fieldlist_NCAR.jsonc [hunk 6]
```diff
@@ -89,12 +89,12 @@
       "units": "m s-1",
       "ndim": 3
     },
-    "TREFHT" : {
+    // "TREFHT" : {
```
Comment (wrongkindofdoctor) [inline data/fieldlist_NCAR.jsonc:92]: Why are TREFHT and T commented out?

**Review by wrongkindofdoctor — CHANGES_REQUESTED** [review]
Please address questions and suggestions in the file comments.

File: data/fieldlist_NCAR.jsonc [hunk 7]
```diff
@@ -90,7 +90,7 @@
       "ndim": 3
     },
     "TREFHT" : {
-      // correct name? CMIP6 equivalent should be tas, temp at 2m ref height
+      correct name? CMIP6 equivalent should be tas, temp at 2m ref height
```
Comment (wrongkindofdoctor) [inline data/fieldlist_NCAR.jsonc:93]: Add "//" back to this line

**Comment by wrongkindofdoctor** [comment]
@drewmresnick It looks like you accidentally uncommented a note in the NCAR fieldlist file that is causing the CI to fail.

Also, LGTM identified a few unused variables and imports that should be removed once you review the output. If you can also address the except block handling the base exception issue, that is good, but not required.

**Comment by drewmresnick** [comment]
> @drewmresnick It looks like you accidentally uncommented a note in the NCAR fieldlist file that is causing the CI to fail.
> 
> Also, LGTM identified a few unused variables and imports that should be removed once you review the output. If you can fix the except block handling the base exception issue, that is good, but not required.

Okay I am looking into these now.

**Comment by drewmresnick** [comment]
@wrongkindofdoctor I also notice now that when I try to run my POD it exits with a key error..

`ERROR: Unrecognized variable name convention 'CESM'`

Is there a quick fix you know of for this error?

File: data/fieldlist_NCAR.jsonc [hunk 8]
```diff
@@ -189,7 +199,7 @@
       "standard_name": "air_temperature",
       "units": "K",
       "ndim": 4
-    },
+    // },
```
Comment (wrongkindofdoctor) [inline data/fieldlist_NCAR.jsonc:202]: @drewmresnick Un-comment this closing bracket--it's the cause of the json parsing error in the latest CI run.

File: src/install.py [hunk 9]
```diff
@@ -335,7 +335,7 @@ def parse_config(self):
         if d.env_setup == 'conda-basic':
             d.conda_envs = [self.settings.conda['framework_env'], 'NCL_base']
             d.environment_manager = "Conda"
-            d.pods = ["Wheeler_Kiladis", "EOF_500hPa", "MJO_suite", "MJO_teleconnection"]
+            d.pods = ["Wheeler_Kiladis", "EOF_500hPa", "MJO_suite", "MJO_teleconnection", "flow_dep_diag"]
```
Comment (wrongkindofdoctor) [inline src/install.py:338]: You shouldn't need to modify anything in the src files (unless @tsjackson-noaa or @jkrasting recommended it). The basic setup applies only to the 1st-gen PODs that run NCL.

File: src/default_tests.jsonc [hunk 10]
```diff
@@ -16,17 +16,18 @@
     // uncomment the second to run NOAA-GFDL-AM4 only for the MJO_prop_amp POD,
     // and likewise for the SM_ET_coupling POD.
     {
-      "CASENAME" : "QBOi.EXP1.AMIP.001",
+      "CASENAME" : "QBOi_CROPPED.EXP1.AMIP.001",
```
Comment (wrongkindofdoctor) [inline src/default_tests.jsonc:19]: Remove the changes from the default_tests.jsonc file. We can eventually include the POD in the file, but we need to make sure that we have the  test dataset available on our FTP server, and need to place QBOi_CROPPED in its own case block.

File: data/fieldlist_NCAR.jsonc (hunk 8 above)
Comment (wrongkindofdoctor) [reply data/fieldlist_NCAR.jsonc]: @drewmresnick The missing NCAR key error still seems to occur because of fieldlist conflicts.  TREFHT and T250 have the same long names,. Changing the T250 long name to "temperature_at_250_mbar_pressure_surface" (I used the LENS2 [CESM variable output list](https://www.cesm.ucar.edu/projects/community-projects/LENS2/variable-list/) as a guide) or something similar should hopefully fix this issue. I'll see if I can replicate the issue on my machine.

Comment (drewmresnick) [reply data/fieldlist_NCAR.jsonc]: Okay I did what you suggested, and it seems to be working now.

**Review by jkrasting — DISMISSED** [review]

**Comment by drewmresnick** [comment]
@wrongkindofdoctor I am wondering if there are any major fixes needed before this can be merged?

**Comment by wrongkindofdoctor** [comment]
> @wrongkindofdoctor I am wondering if there are any major fixes needed before this can be merged?

You have multiple entries in fieldlist_NCAR.jsonc with the same long name as Z3, which is causing the CI to fail because variable long names must be unique. Renaming the `Z250` standard name to something like "geopotential_height_at_250_mbar" should fix it.

File: diagnostics/flow_dep_diag/flow_dep_diag.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+# Required programming language and libraries
+#
+#   This diagnostic runs on the most recent version of python3.
+#   The required packages are as follows, and all should be the most updated version.
+#   Python Libraries used: "netCDF4", "xarray", "numpy", "sklearn",
+#                          "cartopy", "matplotlib", "numba", "datetime", "typing"
+#
+# Required model output variables
+#
+#   Geopotential height anomalies (units: HPa, daily resolution)
```
Comment (wrongkindofdoctor) [inline diagnostics/flow_dep_diag/flow_dep_diag.py:48]: units: ~~H~~hPa

File: src/default_tests.jsonc [hunk 12]
```diff
@@ -23,10 +23,10 @@
       "LASTYR" : 1981,
       "pod_list": [
           // Optional: PODs to run for this model only (defaults to all)
-          "Wheeler_Kiladis",
-          "EOF_500hPa",
-          "MJO_suite",
-          "MJO_teleconnection"
+          // "Wheeler_Kiladis",
```
Comment (wrongkindofdoctor) [inline src/default_tests.jsonc:26]: uncomment the first 4 PODs so that the default.jsonc file is in its original develop branch state

**Review by wrongkindofdoctor — CHANGES_REQUESTED** [review]
@drewmresnick I've requested some final minor corrections. Also, remove the unused variables and module import caught by LGTM.

**Review by wrongkindofdoctor — APPROVED** [review]

File: diagnostics/flow_dep_diag/ClimAnom_func.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+from datetime import datetime
+np.seterr(divide='ignore', invalid='ignore')
+
+### Function to calculate climatology anomalies for each variable in the diagnostic.
+
+time_var = os.environ["time_coord"] #set environment variable for time equal to var for function
+#function to calculate climatology anomalies for eah variable
+def climAnom(var_path, var_name):
+
+    ds = xr.open_dataset(var_path, decode_times = True)
```
Comment (jkrasting) [inline diagnostics/flow_dep_diag/ClimAnom_func.py:13]:
```suggestion
    ds = xr.open_dataset(var_path, use_cftime=True)
```

File: diagnostics/flow_dep_diag/PyWR.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+import numpy as np
+from numba import jit
+import matplotlib.pyplot as plt
+from sklearn.cluster import KMeans
+from sklearn.decomposition import PCA
+from typing import Tuple
+
+def get_number_eof(X: np.ndarray, var_to_explain: float, plot=False) -> int:
+    """Get the number of EOFs of X that explain a given variance proportion
```
Comment (jkrasting) [inline diagnostics/flow_dep_diag/PyWR.py:18]: Is this a rotated EOF?

File: diagnostics/flow_dep_diag/cropping.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+#lon in range -180,180
+nla=71.36
+sla=18.91
+wlo=-171.79
+elo=-66.96
+ds = xr.open_dataset(file_path, decode_times=False)
+
+#the diagnostic assumes your grid has a longitudinal range -180,180
+#Shift your grid from 0,360 using the following lines
+ds.coords['lon'] = (ds.coords['lon'] + 180) % 360 - 180 #shift the values
```
Comment (jkrasting) [inline diagnostics/flow_dep_diag/cropping.py:16]: Test condition before applying the shift.

File: diagnostics/flow_dep_diag/flow_dep_diag.yml [hunk 16]
```diff
@@ -0,0 +1,103 @@
+name: MDTF_flow_dep_diag
```
Comment (jkrasting) [inline diagnostics/flow_dep_diag/flow_dep_diag.yml:1]: This YAML is overly prescriptive.  At a minimum, remove the build codes.  Consider just listing the main top-level packages rather than every dependency.

**Review by jkrasting — CHANGES_REQUESTED** [review]

File: data/fieldlist_NCAR.jsonc [hunk 17]
```diff
@@ -157,6 +156,22 @@
       "units": "W m-2",
       "ndim": 3
     },
+    // Variables for flow_dep_diag module:
```
[the remark this answers is not in the document]
Comment (drewmresnick) [reply data/fieldlist_NCAR.jsonc:159]: @jkrasting Is this then not something I need to worry about?

[the remark this answers is not in the document]
Comment (wrongkindofdoctor) [reply data/fieldlist_NCAR.jsonc:159]: @drewmresnick This is a todo for the framework team, so one of us will take care of it.

[the remark this answers is not in the document]
Comment (tsjackson-noaa) [reply data/fieldlist_NCAR.jsonc:159]: Thanks for bringing this to my attention. Unfortunately constructions of the from `variable_at_XX_mbar` aren't CF standard names [*]. 

The way variables on pressure levels are intended to be specified in the settings file is described in the [documentation](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/ref_settings.html#varlist-entry-properties): the level is specified separately with a `scalar_coordinate` attribute, which refers to a vertical coordinate axis defined in the [dimensions](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/ref_settings.html#dimensions-section) section of the file. By spelling out what's meant in more detail like this, it e.g. enables the framework to extract levels from 3D data on the fly in a fully units-aware way. 

For more examples of how this is set up, look at the [settings file](https://github.com/NOAA-GFDL/MDTF-diagnostics/blob/a7b13c2311ca7af10dc89f00a35d6e50c3903a91/diagnostics/Wheeler_Kiladis/settings.jsonc#L54-L64) for the Wheeler-Kiladis diagnostic.

[*]: I haven't been able to find a concise reference for this: the upshot is that standard names are meant to identify [physical quantities](http://cfconventions.org/cf-conventions/cf-conventions.html#standard-name), independently of how that quantity is sampled in space or time (but there are exceptions to this, such as `air_pressure_at_sea_level`, which deal with physical rather than mathematical boundaries). Variables on pressure levels in the CMIP6 data request don't use this construction, and pressure levels aren't one of the cases covered in [construction of derived standard names](https://cfconventions.org/Data/cf-standard-names/docs/guidelines.html).

File: diagnostics/flow_dep_diag/flow_dep_diag.yml (hunk 16 above)
Comment (drewmresnick) [reply diagnostics/flow_dep_diag/flow_dep_diag.yml:1]: @jkrasting  Is there a way to automatically generate the yaml file with only top-level packages. I know at least for removing the build codes I can use the --no-builds flag.

File: diagnostics/flow_dep_diag/cropping.py (hunk 15 above)
Comment (drewmresnick) [reply diagnostics/flow_dep_diag/cropping.py:16]: I do not think it needs to test for a condition because if the values are between -180,180 the correct grid units will be retained using this calculation. I will add comments to explain this.

File: diagnostics/flow_dep_diag/PyWR.py (hunk 14 above)
Comment (drewmresnick) [reply diagnostics/flow_dep_diag/PyWR.py:18]: it is not a rotated EOF.

File: diagnostics/flow_dep_diag/flow_dep_diag.yml (hunk 16 above)
Comment (wrongkindofdoctor) [reply diagnostics/flow_dep_diag/flow_dep_diag.yml:1]: @drewmresnick you can try using `conda env export --from-history flow_dep_diag.yml` per [this SO reply](https://stackoverflow.com/a/58487556). However, the post states that this will not include package versions. Otherwise, you'll have to modify the file manually, as conda does not seem to have this exact capability.

**Comment by ahmedfiaz** [comment]
Hi @drewmresnick, just checking if you are still able to contribute to this PR? It looked like your PR was quite close to being merged.
